### PR TITLE
harden REST API `atSlot` against non-finalized blocks

### PR DESCRIPTION
**Pull request by arnetheduck**
* harden validator API against pre-finalized slot requests
* check `syncHorizon` when responding to validator api requests too far
from `head`
* limit state-id based requests to one epoch ahead of `head`
* put historic data bounds on block/attestation/etc validator production API, preventing them from being used with already-finalized slots
* add validator block smoke tests
* make rest test create a new genesis with the tests running roughly in
the first epoch to allow testing a few more boundary conditions

### CODE REVIEW

File: beacon_chain/rpc/rest_utils.nim [hunk 1]
```diff
@@ -39,20 +40,13 @@ proc validate(key: string, value: string): int =
...
-    err("Requesting slot too far ahead of the current head")
-
 proc getCurrentHead*(node: BeaconNode, slot: Slot): Result[BlockRef, cstring] =
-  let res = node.dag.head
-  # if not(node.isSynced(res)):
-  #   return err("Cannot fulfill request until node is synced")
-  if res.slot + uint64(2 * SLOTS_PER_EPOCH) < slot:
+  let head = node.dag.head
+
+  if slot > head.slot and not node.isSynced(head):
```
Comment (etan-status) [inline beacon_chain/rpc/rest_utils.nim:46]: `node.isSynced` returns `false` when `head` is more than `syncHorizon` behind the wall clock, it does not take into account the requested `slot`.

The old logic allowed `slot` to be `syncHorizon` ahead of `head` even while not yet fully synced.
The new logic no longer allows even requesting a single slot ahead of `head` while sync has not completed.

Intended?

File: beacon_chain/rpc/rest_utils.nim [hunk 2]
```diff
@@ -64,7 +58,13 @@ proc getBlockSlotId*(node: BeaconNode,
                      stateIdent: StateIdent): Result[BlockSlotId, cstring] =
   case stateIdent.kind
   of StateQueryKind.Slot:
-    let bsi = node.dag.getBlockIdAtSlot(? node.getCurrentSlot(stateIdent.slot)).valueOr:
+    # Limit requests by state id to the next epoch with respect to the current
+    # head to avoid long empty slot replays (in particular a second epoch
+    # transition)
+    if stateIdent.slot.epoch + 1 > node.dag.head.slot.epoch:
+      return err("Requesting state too far ahead of head current head")
```
Comment (etan-status) [inline beacon_chain/rpc/rest_utils.nim]: - There is an extra `head` in the error message.
- Logic looks alright, `slot.epoch + 1` does not seem to be overflowable.

File: beacon_chain/rpc/rest_utils.nim [hunk 3]
```diff
@@ -84,8 +84,12 @@ proc getBlockSlotId*(node: BeaconNode,
...
       ok(node.dag.finalizedHead.toBlockSlotId().expect("not nil"))
     of StateIdentType.Justified:
-      ok(node.dag.head.atEpochStart(getStateField(
-        node.dag.headState, current_justified_checkpoint).epoch).toBlockSlotId().expect("not nil"))
+      # Take checkpoint-synced nodes into account
+      let justifiedEpoch =
+        max(
+          getStateField(node.dag.headState, current_justified_checkpoint).epoch,
+          node.dag.finalizedHead.slot.epoch)
+      ok(node.dag.head.atEpochStart(justifiedEpoch).toBlockSlotId().expect("not nil"))
```
Comment (etan-status) [inline beacon_chain/rpc/rest_utils.nim:92]: Good catch!

File: beacon_chain/rpc/rest_validator_api.nim [hunk 4]
```diff
@@ -606,19 +615,22 @@ proc installValidatorApiHandlers*(router: var RestRouter, node: BeaconNode) =
...
-        let rslot = res.get()
-        if epoch(rslot) < node.dag.cfg.ALTAIR_FORK_EPOCH:
-          return RestApiResponse.jsonError(Http400,
-                                           SlotFromTheIncorrectForkError)
-        rslot
+
+      let res = slot.get()
+      if res.isErr():
+        return RestApiResponse.jsonError(Http400, InvalidSlotValueError,
+                                          $res.error())
```
Comment (etan-status) [inline beacon_chain/rpc/rest_validator_api.nim:647]: NIT: Indentation of the second line w.r.t. `(`, on all of these.

File: beacon_chain/rpc/rest_validator_api.nim [hunk 5]
```diff
@@ -400,6 +406,9 @@ proc installValidatorApiHandlers*(router: var RestRouter, node: BeaconNode) =
               return RestApiResponse.jsonError(Http400, InvalidSlotValueError,
                                                $res.error())
             res.get()
+        if qslot <= node.dag.finalizedHead.slot:
```
Comment (etan-status) [inline beacon_chain/rpc/rest_validator_api.nim:420]: The altair check of other endpoints is inside the block fetching `qslot` -- should this additional check be located similarly, as in, first loading into `rslot`, then validating against the allowed range, then assigning to `qslot`? It doesn't matter semantically, it's more of a style / consistency thing.

**Review by etan-status — APPROVED** [review]

File: beacon_chain/rpc/rest_utils.nim (hunk 1 above)
Comment (arnetheduck) [reply beacon_chain/rpc/rest_utils.nim:46]: yes, while syncing we don't want to allow spurious requests for slots that will cause expensive empty-slots replays of the state

File: beacon_chain/rpc/rest_utils.nim (hunk 2 above)
Comment (arnetheduck) [reply beacon_chain/rpc/rest_utils.nim]: actually it was overflowing because of the magic far_future_epoch handling (that at some point should be applied to `+` as well...)

File: beacon_chain/rpc/rest_validator_api.nim (hunk 5 above)
Comment (arnetheduck) [reply beacon_chain/rpc/rest_validator_api.nim:420]: indeed, there's a myriad of these little improvements that one _could_ make with infinite time at hand..

File: beacon_chain/rpc/rest_validator_api.nim (hunk 4 above)
Comment (arnetheduck) [reply beacon_chain/rpc/rest_validator_api.nim:647]: this is true, but also a bit of a chore / best-effort thing as long as we don't have an  autoformatter.. the best way to highlight  these is through the little suggestion button which gives the right perfomance/effort ratio for fixing them (and avoids typing NIT in ALL-CAPS :))

Screenshot from 2022-03-22 20-17-04
